{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "630cb915-32f0-4fe9-93d4-06b244fc62ae",
   "metadata": {},
   "source": [
    "# DeMetRA - review figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1adeee0b-883f-49c1-b0ad-9b6df095e6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae830b76-d455-4668-987f-25dc3f2d9c40",
   "metadata": {},
   "source": [
    "## Is-processed dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e1e90d31-2812-4f80-bd61-44de1710431a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Type</th>\n",
       "      <th>Author_dirty</th>\n",
       "      <th>Year</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal_dirty</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Include</th>\n",
       "      <th>pdf</th>\n",
       "      <th>Sample_size_total</th>\n",
       "      <th>...</th>\n",
       "      <th>Missing_value_note</th>\n",
       "      <th>Reflect_phenotype</th>\n",
       "      <th>Link</th>\n",
       "      <th>Author_list</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Date</th>\n",
       "      <th>URL</th>\n",
       "      <th>Author</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Abrishamcar (2022) - DNA methylation as a pote...</td>\n",
       "      <td>262</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | Explained variance | R2 = 0.23</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3_x...</td>\n",
       "      <td>['Abrishamcar, S.', 'Chen, J.', 'Feil, D.', 'K...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prenatal tobacco exposure (PTE) and prenatal a...</td>\n",
       "      <td>2022-09-30</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3</td>\n",
       "      <td>Abrishamcar et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Abrishamcar (2022) - DNA methylation as a pote...</td>\n",
       "      <td>262</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | Explained variance | R = 0.0006</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3_x...</td>\n",
       "      <td>['Abrishamcar, S.', 'Chen, J.', 'Feil, D.', 'K...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prenatal tobacco exposure (PTE) and prenatal a...</td>\n",
       "      <td>2022-09-30</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3</td>\n",
       "      <td>Abrishamcar et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>R. Al-Jawahiri; A. Foroutan; J. Kerkhof; H. Mc...</td>\n",
       "      <td>2022</td>\n",
       "      <td>SOX11 variants cause a neurodevelopmental diso...</td>\n",
       "      <td>Gen Med</td>\n",
       "      <td>10.1016/j.gim.2022.02.013</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Al-Jawahiri (2022) - SOX11 variants cause a ne...</td>\n",
       "      <td>126</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No actual number reported</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>['Al-Jawahiri, R.', 'Foroutan, A.', 'Kerkhof, ...</td>\n",
       "      <td>Gen Med</td>\n",
       "      <td>['nucleic acid genotyping array kit', 'growth ...</td>\n",
       "      <td>Purpose: This study aimed to undertake a multi...</td>\n",
       "      <td>2022-04-05</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Al-Jawahiri et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>R. Alfano; D. Zugna; H. Barros; M. Bustamante;...</td>\n",
       "      <td>2023</td>\n",
       "      <td>Cord blood epigenome-wide meta-analysis in six...</td>\n",
       "      <td>BMC Med</td>\n",
       "      <td>10.1186/s12916-022-02685-7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Alfano (2023) - Cord blood epigenome-wide meta</td>\n",
       "      <td>2003</td>\n",
       "      <td>...</td>\n",
       "      <td>Imputed | missForest</td>\n",
       "      <td>Reported | AUC | 0.61 (95% CI= 0.55–0.68)</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>['Alfano, R.', 'Zugna, D.', 'Barros, H.', 'Bus...</td>\n",
       "      <td>BMC Med</td>\n",
       "      <td>['aurora C kinase', 'long untranslated RNA', '...</td>\n",
       "      <td>Background: Rapid postnatal growth may result ...</td>\n",
       "      <td>2023-01-17</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Alfano et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>E. Aref-Eshghi; E. G. Bend; S. Colaiacovo; M. ...</td>\n",
       "      <td>2019</td>\n",
       "      <td>Diagnostic Utility of Genome-wide DNA Methylat...</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>10.1016/j.ajhg.2019.03.008</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Aref-Eshghi (2019) - Diagnostic Utility of Gen...</td>\n",
       "      <td>575</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No actual number reported</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>['Aref-Eshghi, E.', 'Bend, E. G.', 'Colaiacovo...</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>['genomic DNA', 'adult', 'article', 'case repo...</td>\n",
       "      <td>Conventional genetic testing of individuals wi...</td>\n",
       "      <td>2019-04-05</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Aref-Eshghi et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>190</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Kallak, T. K., Fransson, E., Bränn, E., Berglu...</td>\n",
       "      <td>2022</td>\n",
       "      <td>Maternal prenatal depressive symptoms and todd...</td>\n",
       "      <td>Translation psychiatry</td>\n",
       "      <td>10.1038/s41398-022-01954-6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Kallak (2022) - Maternal prenatal depressive s...</td>\n",
       "      <td>256</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | ranged between .4-.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Children of mothers with prenatal depressive s...</td>\n",
       "      <td>2022-05-05</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-01954-6</td>\n",
       "      <td>Kallak et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>190</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Kallak, T. K., Fransson, E., Bränn, E., Berglu...</td>\n",
       "      <td>2022</td>\n",
       "      <td>Maternal prenatal depressive symptoms and todd...</td>\n",
       "      <td>Translation psychiatry</td>\n",
       "      <td>10.1038/s41398-022-01954-6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Kallak (2022) - Maternal prenatal depressive s...</td>\n",
       "      <td>256</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | ranged between .4-.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Children of mothers with prenatal depressive s...</td>\n",
       "      <td>2022-05-05</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-01954-6</td>\n",
       "      <td>Kallak et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>190</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Kallak, T. K., Fransson, E., Bränn, E., Berglu...</td>\n",
       "      <td>2022</td>\n",
       "      <td>Maternal prenatal depressive symptoms and todd...</td>\n",
       "      <td>Translation psychiatry</td>\n",
       "      <td>10.1038/s41398-022-01954-6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Kallak (2022) - Maternal prenatal depressive s...</td>\n",
       "      <td>256</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | ranged between .4-.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Children of mothers with prenatal depressive s...</td>\n",
       "      <td>2022-05-05</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-01954-6</td>\n",
       "      <td>Kallak et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>190</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Kallak, T. K., Fransson, E., Bränn, E., Berglu...</td>\n",
       "      <td>2022</td>\n",
       "      <td>Maternal prenatal depressive symptoms and todd...</td>\n",
       "      <td>Translation psychiatry</td>\n",
       "      <td>10.1038/s41398-022-01954-6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Kallak (2022) - Maternal prenatal depressive s...</td>\n",
       "      <td>256</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | ranged between .4-.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Children of mothers with prenatal depressive s...</td>\n",
       "      <td>2022-05-05</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-01954-6</td>\n",
       "      <td>Kallak et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>191</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Pishva, E., van den Hove, D. L., Laroche, V., ...</td>\n",
       "      <td>2023</td>\n",
       "      <td>Genome-wide DNA methylation analysis of aggres...</td>\n",
       "      <td>The Journal of Child Psychology and Psychiatry</td>\n",
       "      <td>10.1111/jcpp.13782</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Pishva (2023) - Genome-wide DNA methylation an...</td>\n",
       "      <td>95</td>\n",
       "      <td>...</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | AUC =  0.89 [CI = 0.77-1.00]; 87% s...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Pishva, E.', 'van den Hove, D. L. A.', 'Laro...</td>\n",
       "      <td>J Child Psychol Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BACKGROUND: Human aggression is influenced by ...</td>\n",
       "      <td>2023-07-11</td>\n",
       "      <td>http://dx.doi.org/10.1111/jcpp.13782</td>\n",
       "      <td>Pishva et al.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>760 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Identifier             Type  \\\n",
       "0             1  Journal Article   \n",
       "1             1  Journal Article   \n",
       "2             3  Journal Article   \n",
       "3             4  Journal Article   \n",
       "4             7  Journal Article   \n",
       "..          ...              ...   \n",
       "755         190  Journal Article   \n",
       "756         190  Journal Article   \n",
       "757         190  Journal Article   \n",
       "758         190  Journal Article   \n",
       "759         191  Journal Article   \n",
       "\n",
       "                                          Author_dirty  Year  \\\n",
       "0    S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...  2022   \n",
       "1    S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...  2022   \n",
       "2    R. Al-Jawahiri; A. Foroutan; J. Kerkhof; H. Mc...  2022   \n",
       "3    R. Alfano; D. Zugna; H. Barros; M. Bustamante;...  2023   \n",
       "4    E. Aref-Eshghi; E. G. Bend; S. Colaiacovo; M. ...  2019   \n",
       "..                                                 ...   ...   \n",
       "755  Kallak, T. K., Fransson, E., Bränn, E., Berglu...  2022   \n",
       "756  Kallak, T. K., Fransson, E., Bränn, E., Berglu...  2022   \n",
       "757  Kallak, T. K., Fransson, E., Bränn, E., Berglu...  2022   \n",
       "758  Kallak, T. K., Fransson, E., Bränn, E., Berglu...  2022   \n",
       "759  Pishva, E., van den Hove, D. L., Laroche, V., ...  2023   \n",
       "\n",
       "                                                 Title  \\\n",
       "0    DNA methylation as a potential mediator of the...   \n",
       "1    DNA methylation as a potential mediator of the...   \n",
       "2    SOX11 variants cause a neurodevelopmental diso...   \n",
       "3    Cord blood epigenome-wide meta-analysis in six...   \n",
       "4    Diagnostic Utility of Genome-wide DNA Methylat...   \n",
       "..                                                 ...   \n",
       "755  Maternal prenatal depressive symptoms and todd...   \n",
       "756  Maternal prenatal depressive symptoms and todd...   \n",
       "757  Maternal prenatal depressive symptoms and todd...   \n",
       "758  Maternal prenatal depressive symptoms and todd...   \n",
       "759  Genome-wide DNA methylation analysis of aggres...   \n",
       "\n",
       "                                      Journal_dirty  \\\n",
       "0                                 Transl Psychiatry   \n",
       "1                                 Transl Psychiatry   \n",
       "2                                           Gen Med   \n",
       "3                                           BMC Med   \n",
       "4                                    Am J Hum Genet   \n",
       "..                                              ...   \n",
       "755                          Translation psychiatry   \n",
       "756                          Translation psychiatry   \n",
       "757                          Translation psychiatry   \n",
       "758                          Translation psychiatry   \n",
       "759  The Journal of Child Psychology and Psychiatry   \n",
       "\n",
       "                            DOI Include  \\\n",
       "0    10.1038/s41398-022-02195-3     Yes   \n",
       "1    10.1038/s41398-022-02195-3     Yes   \n",
       "2     10.1016/j.gim.2022.02.013     Yes   \n",
       "3    10.1186/s12916-022-02685-7     Yes   \n",
       "4    10.1016/j.ajhg.2019.03.008     Yes   \n",
       "..                          ...     ...   \n",
       "755  10.1038/s41398-022-01954-6     Yes   \n",
       "756  10.1038/s41398-022-01954-6     Yes   \n",
       "757  10.1038/s41398-022-01954-6     Yes   \n",
       "758  10.1038/s41398-022-01954-6     Yes   \n",
       "759          10.1111/jcpp.13782     Yes   \n",
       "\n",
       "                                                   pdf Sample_size_total  ...  \\\n",
       "0    Abrishamcar (2022) - DNA methylation as a pote...               262  ...   \n",
       "1    Abrishamcar (2022) - DNA methylation as a pote...               262  ...   \n",
       "2    Al-Jawahiri (2022) - SOX11 variants cause a ne...               126  ...   \n",
       "3       Alfano (2023) - Cord blood epigenome-wide meta              2003  ...   \n",
       "4    Aref-Eshghi (2019) - Diagnostic Utility of Gen...               575  ...   \n",
       "..                                                 ...               ...  ...   \n",
       "755  Kallak (2022) - Maternal prenatal depressive s...               256  ...   \n",
       "756  Kallak (2022) - Maternal prenatal depressive s...               256  ...   \n",
       "757  Kallak (2022) - Maternal prenatal depressive s...               256  ...   \n",
       "758  Kallak (2022) - Maternal prenatal depressive s...               256  ...   \n",
       "759  Pishva (2023) - Genome-wide DNA methylation an...                95  ...   \n",
       "\n",
       "       Missing_value_note                                  Reflect_phenotype  \\\n",
       "0            Not reported          Reported | Explained variance | R2 = 0.23   \n",
       "1            Not reported         Reported | Explained variance | R = 0.0006   \n",
       "2                      No                          No actual number reported   \n",
       "3    Imputed | missForest          Reported | AUC | 0.61 (95% CI= 0.55–0.68)   \n",
       "4                      No                          No actual number reported   \n",
       "..                    ...                                                ...   \n",
       "755          Not reported                    Reported | ranged between .4-.6   \n",
       "756          Not reported                    Reported | ranged between .4-.6   \n",
       "757          Not reported                    Reported | ranged between .4-.6   \n",
       "758          Not reported                    Reported | ranged between .4-.6   \n",
       "759          Not reported  Reported | AUC =  0.89 [CI = 0.77-1.00]; 87% s...   \n",
       "\n",
       "                                                  Link  \\\n",
       "0    http://dx.doi.org/10.1038/s41398-022-02195-3_x...   \n",
       "1    http://dx.doi.org/10.1038/s41398-022-02195-3_x...   \n",
       "2    https://www.embase.com/search/results?subactio...   \n",
       "3    https://www.embase.com/search/results?subactio...   \n",
       "4    https://www.embase.com/search/results?subactio...   \n",
       "..                                                 ...   \n",
       "755                                                NaN   \n",
       "756                                                NaN   \n",
       "757                                                NaN   \n",
       "758                                                NaN   \n",
       "759                                                NaN   \n",
       "\n",
       "                                           Author_list  \\\n",
       "0    ['Abrishamcar, S.', 'Chen, J.', 'Feil, D.', 'K...   \n",
       "1    ['Abrishamcar, S.', 'Chen, J.', 'Feil, D.', 'K...   \n",
       "2    ['Al-Jawahiri, R.', 'Foroutan, A.', 'Kerkhof, ...   \n",
       "3    ['Alfano, R.', 'Zugna, D.', 'Barros, H.', 'Bus...   \n",
       "4    ['Aref-Eshghi, E.', 'Bend, E. G.', 'Colaiacovo...   \n",
       "..                                                 ...   \n",
       "755  ['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...   \n",
       "756  ['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...   \n",
       "757  ['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...   \n",
       "758  ['Kallak, T. K.', 'Fransson, E.', 'Brann, E.',...   \n",
       "759  ['Pishva, E.', 'van den Hove, D. L. A.', 'Laro...   \n",
       "\n",
       "                        Journal  \\\n",
       "0             Transl Psychiatry   \n",
       "1             Transl Psychiatry   \n",
       "2                       Gen Med   \n",
       "3                       BMC Med   \n",
       "4                Am J Hum Genet   \n",
       "..                          ...   \n",
       "755           Transl Psychiatry   \n",
       "756           Transl Psychiatry   \n",
       "757           Transl Psychiatry   \n",
       "758           Transl Psychiatry   \n",
       "759  J Child Psychol Psychiatry   \n",
       "\n",
       "                                              Keywords  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2    ['nucleic acid genotyping array kit', 'growth ...   \n",
       "3    ['aurora C kinase', 'long untranslated RNA', '...   \n",
       "4    ['genomic DNA', 'adult', 'article', 'case repo...   \n",
       "..                                                 ...   \n",
       "755                                                NaN   \n",
       "756                                                NaN   \n",
       "757                                                NaN   \n",
       "758                                                NaN   \n",
       "759                                                NaN   \n",
       "\n",
       "                                              Abstract        Date  \\\n",
       "0    Prenatal tobacco exposure (PTE) and prenatal a...  2022-09-30   \n",
       "1    Prenatal tobacco exposure (PTE) and prenatal a...  2022-09-30   \n",
       "2    Purpose: This study aimed to undertake a multi...  2022-04-05   \n",
       "3    Background: Rapid postnatal growth may result ...  2023-01-17   \n",
       "4    Conventional genetic testing of individuals wi...  2019-04-05   \n",
       "..                                                 ...         ...   \n",
       "755  Children of mothers with prenatal depressive s...  2022-05-05   \n",
       "756  Children of mothers with prenatal depressive s...  2022-05-05   \n",
       "757  Children of mothers with prenatal depressive s...  2022-05-05   \n",
       "758  Children of mothers with prenatal depressive s...  2022-05-05   \n",
       "759  BACKGROUND: Human aggression is influenced by ...  2023-07-11   \n",
       "\n",
       "                                                   URL              Author  \n",
       "0         http://dx.doi.org/10.1038/s41398-022-02195-3  Abrishamcar et al.  \n",
       "1         http://dx.doi.org/10.1038/s41398-022-02195-3  Abrishamcar et al.  \n",
       "2    https://www.embase.com/search/results?subactio...  Al-Jawahiri et al.  \n",
       "3    https://www.embase.com/search/results?subactio...       Alfano et al.  \n",
       "4    https://www.embase.com/search/results?subactio...  Aref-Eshghi et al.  \n",
       "..                                                 ...                 ...  \n",
       "755       http://dx.doi.org/10.1038/s41398-022-01954-6       Kallak et al.  \n",
       "756       http://dx.doi.org/10.1038/s41398-022-01954-6       Kallak et al.  \n",
       "757       http://dx.doi.org/10.1038/s41398-022-01954-6       Kallak et al.  \n",
       "758       http://dx.doi.org/10.1038/s41398-022-01954-6       Kallak et al.  \n",
       "759               http://dx.doi.org/10.1111/jcpp.13782       Pishva et al.  \n",
       "\n",
       "[760 rows x 47 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assets_directory = '../assets/'\n",
    "\n",
    "lit = pd.read_csv(f'{assets_directory}MPS_literature_cleaned.csv')\n",
    "d = pd.read_csv(f'{assets_directory}MPS_base_target.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23fbf98f-ec94-44d1-9e4d-4d248964c18e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "demetra_pallete = {\n",
    "    'caribbean-current': '#00635dff',\n",
    "    'vista-blue': '#809bceff',\n",
    "    'xanthous': '#ffc145ff',\n",
    "    'fairy-tale': '#eac4d5ff',\n",
    "    'red-ncs': '#b20d30ff',\n",
    "    'jasmine': '#ffd685ff',\n",
    "    'yinmn-blue': '#395993ff',\n",
    "    'light-sea-green': '#00b8abff',\n",
    "    'quinacridone-magenta': '#873157ff',\n",
    "    'lavender-blush': '#feecf0ff'}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d04ed1bb",
   "metadata": {},
   "source": [
    "### Pie charts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d1150f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pie_count(ax, var, df=d):\n",
    "\n",
    "    count = df.value_counts(var, dropna=False)\n",
    "    labels = count.index\n",
    "    sizes = count\n",
    "    \n",
    "    ax.pie(sizes, labels=labels, autopct='%1.1f%%', colors=demetra_pallete.values())\n",
    "\n",
    "\n",
    "def base_target_overlap(var, df=d):\n",
    "    ovlp = df[f'{var}_base'] == df[f'{var}_targ']\n",
    "    return(f'{round((sum(ovlp) / len(ovlp))*100)}%')\n",
    "\n",
    "\n",
    "def compare_base_target(var_list, df=d):\n",
    "\n",
    "    fs = dict(big=30, small=15)\n",
    "\n",
    "    row_targ = list()\n",
    "    row_base = list()\n",
    "    row_ovlp = list()\n",
    "\n",
    "    for i in range(len(var_list)+1):  \n",
    "        row_targ.append(f'A{i}')\n",
    "        row_base.append(f'B{i}')\n",
    "        row_ovlp.append(f'C{i}')\n",
    "    \n",
    "    fig, axs = plt.subplot_mosaic([row_targ, row_base, row_ovlp], figsize=(25, 13), \n",
    "                                  height_ratios=[1,1,.5], width_ratios=[.2]+[1]*len(var_list))\n",
    "    \n",
    "\n",
    "    label_txt = dict(x=.5, y=.5, ha='center', va='center', fontsize=fs['big'], rotation=90)\n",
    "    \n",
    "    axs['A0'].text(s='Target', transform=axs['A0'].transAxes, **label_txt)\n",
    "    axs['B0'].text(s='Base', transform=axs['B0'].transAxes, **label_txt)\n",
    "    axs['C0'].text(s='Match', transform=axs['C0'].transAxes, **label_txt)\n",
    "\n",
    "    [axs[i].axis('off') for i in ['A0', 'B0', 'C0']]\n",
    "\n",
    "    for i, var in enumerate(var_list):\n",
    "        \n",
    "        pie_count(axs[f'A{i+1}'], f'{var}_targ')\n",
    "        pie_count(axs[f'B{i+1}'], f'{var}_base')\n",
    "\n",
    "        axs[f'C{i+1}'].text(x=.5, y=.5, s=base_target_overlap(var), fontsize=fs['big'], ha='center', va='center', \n",
    "                           transform=axs[f'C{i+1}'].transAxes)\n",
    "        axs[f'C{i+1}'].axis('off')\n",
    "\n",
    "        axs[f'A{i+1}'].set_title('\\n'.join(var.split('_')), fontsize=fs['big'], fontweight='bold') \n",
    "\n",
    "\n",
    "compare_base_target(var_list = ['Tissue', 'Array','Ancestry','Developmental_period'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be350b54",
   "metadata": {},
   "source": [
    "### Sankey plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fcbaa8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sankey(ax, var, left_labels, right_labels, d=d, left='targ', right='base',\n",
    "           title_left='Target', title_right='Base', spacer=10, fss={'sm': 14, 'l': 15, 'xl': 25}):\n",
    "    \n",
    "    counts = pd.DataFrame(d[[f'{var}_{left}',f'{var}_{right}']].value_counts(dropna=False)).reset_index()\n",
    "    \n",
    "    total = counts['count'].sum()\n",
    "    \n",
    "    def size_esimator(label_dict, side):\n",
    "    \n",
    "        size_list = list()\n",
    "        \n",
    "        for label in label_dict.keys():\n",
    "            label_count = int(counts.loc[counts[f'{var}_{side}']==label, 'count'].sum())\n",
    "            label_dict[label]['size'] = label_count\n",
    "            size_list.append(label_count)\n",
    "    \n",
    "        cumulative_sum = np.cumsum(size_list).tolist()\n",
    "        \n",
    "        top_pos = [0] + [c+(spacer*(i+1))for i,c in enumerate(cumulative_sum[:-1])]\n",
    "        bottom_pos = [c+(spacer*(i))for i,c in enumerate(cumulative_sum)]\n",
    "    \n",
    "        for i, label in enumerate(label_dict.keys()):\n",
    "            label_dict[label]['top'] = top_pos[i]\n",
    "            label_dict[label]['bottom'] = bottom_pos[i]\n",
    "        \n",
    "        return label_dict\n",
    "    \n",
    "    left_dict = size_esimator(left_labels, side=left)\n",
    "    right_dict = size_esimator(right_labels, side=right)\n",
    "    \n",
    "    def label_y(label_dict, label):\n",
    "        if label_dict[label]['size'] > spacer:\n",
    "            y = label_dict[label]['top']+1\n",
    "            va='top'\n",
    "        else:\n",
    "            y = label_dict[label]['top'] + label_dict[label]['size']*0.5\n",
    "            va='center'\n",
    "\n",
    "        string_spacer = '\\n' if label_dict[label]['size'] > 5 else ' '\n",
    "        percent_count = round(label_dict[label]['size'] / total * 100)\n",
    "        percent_count = percent_count if percent_count > 0 else '<1'\n",
    "        \n",
    "        s = f\"{label}{string_spacer}({percent_count}%)\"\n",
    "        \n",
    "        return dict(y=y, s=s, va=va)\n",
    "        \n",
    "    # Draw left counts\n",
    "    for label in left_dict.keys():\n",
    "        ax.fill_between(x=[0, 1], y1=left_dict[label]['top'], y2=left_dict[label]['bottom'], \n",
    "                        color=left_dict[label]['color'], edgecolor=None)\n",
    "        ax.text(x=-0.1, **label_y(left_dict, label), ha='right', fontsize=fss['sm'])\n",
    "    \n",
    "    # Draw right counts\n",
    "    for label in right_dict.keys():\n",
    "        ax.fill_between(x=[9, 10], y1=right_dict[label]['top'], y2=right_dict[label]['bottom'], \n",
    "                        color=right_dict[label]['color'], alpha=1, edgecolor=None)\n",
    "        ax.text(x=10.1, **label_y(right_dict, label), ha='left', fontsize=fss['sm'])\n",
    "\n",
    "    # Add titles on each side\n",
    "    titlespecs = dict(y=-10, va='center',ha='center', fontweight='bold', fontsize=fss['l'])\n",
    "    ax.text(x=0.5, s=title_left, **titlespecs)\n",
    "    ax.text(x=9.5, s=title_right, **titlespecs)\n",
    "    \n",
    "    # Draw strips \n",
    "    for left_label in left_dict.keys():\n",
    "        \n",
    "        for right_label in right_dict.keys():\n",
    "            \n",
    "            strip_color = left_dict[left_label]['color'] # Color strip according to the left side\n",
    "            \n",
    "            strip_size = counts.loc[(counts[f'{var}_{left}']==left_label) & (counts[f'{var}_{right}']==right_label), 'count']\n",
    "    \n",
    "            \n",
    "            if  len(strip_size) > 0:\n",
    "                strip_size = int(strip_size.iloc[0])\n",
    "    \n",
    "                # Create array of y values for each strip, half at left value, half at right, convolve\n",
    "                ys_d = np.array(50 * [left_dict[left_label]['top']] + 50 * [right_dict[right_label]['top']])\n",
    "                ys_d = np.convolve(ys_d, 0.05 * np.ones(20), mode='valid')\n",
    "                ys_d = np.convolve(ys_d, 0.05 * np.ones(20), mode='valid')\n",
    "                \n",
    "                ys_u = np.array(50 * [left_dict[left_label]['top'] + strip_size] + 50 * [right_dict[right_label]['top'] + strip_size])\n",
    "                ys_u = np.convolve(ys_u, 0.05 * np.ones(20), mode='valid')\n",
    "                ys_u = np.convolve(ys_u, 0.05 * np.ones(20), mode='valid')\n",
    "    \n",
    "                # Update bottom edges at each label so next strip starts at the right place\n",
    "                left_dict[left_label]['top'] += strip_size\n",
    "                right_dict[right_label]['top'] += strip_size\n",
    "                \n",
    "                ax.fill_between(np.linspace(1, 9, len(ys_d)), ys_d, ys_u, alpha=0.4, color=strip_color, edgecolor=None)\n",
    "    \n",
    "    largest_count = max(left_dict[list(left_dict.keys())[-1]]['bottom'], right_dict[list(right_dict.keys())[-1]]['bottom'])\n",
    "    ax.set_xlim(-0.1, 10.1)\n",
    "    ax.set_ylim(-10, largest_count+10)\n",
    "    ax.invert_yaxis()\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Add superior title\n",
    "    ax.set_title(' '.join(var.split('_')), fontweight='bold', fontsize=fss['xl'], pad=25)\n",
    "    \n",
    "    # Also return overall overla \n",
    "    color_counts = counts.copy()\n",
    "    color_counts[f'{var}_{left}'] = [left_labels[i]['color'] for i in counts[f'{var}_{left}']]\n",
    "    color_counts[f'{var}_{right}'] = [right_labels[i]['color'] for i in counts[f'{var}_{right}']]\n",
    "    \n",
    "    match = int(color_counts.loc[(color_counts[f'{var}_{left}'] == color_counts[f'{var}_{right}']) & (color_counts[f'{var}_{left}'] != 'grey'), \n",
    "                'count'].sum())\n",
    "    match_percent = match / total * 100\n",
    "\n",
    "    return match_percent\n",
    "\n",
    "def display_match(ax, match, fs=22): \n",
    "    ax.text(x=.5, y=.95, s=f'Match: {round(match)}%',fontsize=fs, ha='center', va='center', transform=ax.transAxes)\n",
    "    ax.axis('off')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21beeb84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ['Tissue', 'Array','Ancestry','Developmental_period']\n",
    "def get_counts(var):\n",
    "    counts = pd.DataFrame(d[[f'{var}_targ',f'{var}_base']].value_counts(dropna=False)).reset_index()\n",
    "    print(counts[f'{var}_targ'].unique())\n",
    "    print(counts[f'{var}_base'].unique())\n",
    "    return counts\n",
    "\n",
    "get_counts('Developmental_period')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b34c7f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplot_mosaic('AB;ab;CD;cd', figsize=(20, 30),\n",
    "                              height_ratios=[1,.2, 1,.2], gridspec_kw=dict(hspace=0, wspace=0.7))\n",
    "    \n",
    "\n",
    "a = sankey(axs['A'], var='Array', \n",
    "           left_labels = {'450K': {'color': 'darkgreen'}, \n",
    "                      'EPICv1': {'color': 'mediumpurple'},\n",
    "                      'Multiple (450K, EPICv1)': {'color': 'orange'},\n",
    "                      'Multiple (450K, GMEL (~3000 CpGs from EPICv1))': {'color': 'orange'},\n",
    "                      'Multiple (450K, EPICv2)': {'color': 'orange'}},\n",
    "           right_labels = {'450K': {'color': 'darkgreen'}, \n",
    "                       'EPICv1': {'color': 'mediumpurple'}, \n",
    "                       'Multiple (450K, EPICv1)': {'color': 'orange'}})\n",
    "\n",
    "display_match(axs['a'], a)\n",
    "\n",
    "b = sankey(axs['B'], var='Tissue',\n",
    "           left_labels = {'Peripheral blood': {'color':'crimson'},\n",
    "                          'Whole blood': {'color':'crimson'},\n",
    "                          'Dried bloodspot': {'color':'crimson'},\n",
    "                          'Blood-clots': {'color':'crimson'},\n",
    "                          'Cord blood': {'color':'darkred'},\n",
    "                          'Saliva': {'color':'lightblue'},\n",
    "                          'Buccal cells': {'color':'teal'},\n",
    "                          'Tumour cells': {'color':'orange'},\n",
    "                          'Not reported': {'color':'grey'}}, \n",
    "           right_labels = {'Peripheral blood': {'color':'crimson'},\n",
    "                           'Whole blood': {'color':'crimson'},\n",
    "                           'Cord blood': {'color':'darkred'},\n",
    "                           'Multiple (Cord blood, Dried bloodspot)': {'color':'crimson'},\n",
    "                           'Multiple (Cord blood, Whole blood)': {'color':'crimson'},\n",
    "                           'Multiple (Whole blood, HPCs)': {'color':'crimson'},\n",
    "                           'Leukocytes': {'color':'mediumpurple'},\n",
    "                           'Tumour cells': {'color':'orange'}})\n",
    "\n",
    "display_match(axs['b'], b)\n",
    "\n",
    "c = sankey(axs['C'], var='Ancestry',\n",
    "           left_labels = {'White': {'color':'pink'}, \n",
    "                          'European': {'color':'pink'}, \n",
    "                          'Mixed': {'color':'purple'}, \n",
    "                          'Hispanic': {'color': 'orange'},\n",
    "                          'African': {'color':'crimson'},\n",
    "                          'Not reported': {'color': 'grey'}}, \n",
    "           right_labels = {'White': {'color':'pink'},\n",
    "                           'European': {'color':'pink'},\n",
    "                           'Mixed': {'color':'purple'}, \n",
    "                           'Hispanic': {'color': 'orange'},\n",
    "                           'Not reported': {'color':'grey'}})\n",
    "\n",
    "display_match(axs['c'], c)\n",
    "\n",
    "\n",
    "dp = sankey(axs['D'], var='Developmental_period',\n",
    "           left_labels = {'Birth': {'color':'darkblue'}, \n",
    "               'Very early childhood': {'color':'#4132d4'}, \n",
    "               'Mid childhood': {'color':'#7566ff'}, \n",
    "               'Late childhood': {'color':'#beb7ff'}, \n",
    "               'Childhood and adolescence': {'color':'#f0cdff'}, \n",
    "               'Adolescence': {'color':'purple'}},\n",
    "           right_labels = {'Birth': {'color':'darkblue'}, \n",
    "                'Mid childhood': {'color':'#7566ff'}, \n",
    "                'Late childhood': {'color':'#beb7ff'}, \n",
    "                'Childhood': {'color':'blue'},\n",
    "                'Childhood and adolescence': {'color':'#f0cdff'}, \n",
    "                'Birth, Childhood and adolescence': {'color':'#7b07d0'},\n",
    "                'Adolescence': {'color':'purple'},\n",
    "                'Adults':{'color':'teal'},\n",
    "                'Not reported': {'color':'grey'}})\n",
    "\n",
    "display_match(axs['d'], dp)\n",
    "\n",
    "fig.savefig('Figure3.pdf', dpi=400, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "33ddadcb-d9c0-48e8-8391-8050e896eafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_targ.groupby(d_targ.What_is_available).Identifier_base.value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f329f322-aa43-4187-b0ab-e38d7469cce4",
   "metadata": {},
   "source": [
    "# Other figures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7396ee7-3bed-4188-82e1-8142cd5df77f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify colors to use for categorical variables\n",
    "# qprint([\"Category\"])\n",
    "category_color_map = {'Syndrome': 'royalblue',\n",
    "                      'Disease': 'firebrick',\n",
    "                      'Protein': 'darkviolet',\n",
    "                      'Lifestyle': 'gold',\n",
    "                      'Psychiatric': 'teal',\n",
    "                      'Environmental exposure': 'limegreen',\n",
    "                      'Birth outcome': 'lightblue',\n",
    "                      'Tumour': 'pink',\n",
    "                      'Aging phenotype': 'darkgrey'}\n",
    "\n",
    "tissue_color_map = {'Peripheral blood': 'crimson',\n",
    "                    'Whole blood': 'gold',\n",
    "                    'Cord blood': 'brown',\n",
    "                    'Placenta': 'pink',\n",
    "                    'Blood-clots': 'blueviolet',\n",
    "                    'Dried bloodspot': 'indianred',\n",
    "                    'Saliva': 'lightblue',\n",
    "                    'Buccal cells': 'steelblue',\n",
    "                    'Nasal epithelial cells': 'teal',\n",
    "                    'Tumour cells': 'darkgreen',\n",
    "                    'Leukocytes': 'yellow',\n",
    "                    'Leukocytes': 'yellow',\n",
    "                    'Bone marrow granulocytes': 'magenta',\n",
    "                    'Not reported': 'darkgrey'}\n",
    "\n",
    "array_color_map = {'EPICv2': 'darkblue',\n",
    "                   'EPICv1': 'mediumblue',\n",
    "                   '450K, EPICv1': 'seagreen',\n",
    "                   '450k': 'orange',\n",
    "                   '450K, GMEL (~3000 CpGs from EPICv1)':'orange',\n",
    "                   'WGBS': 'magenta'}\n",
    "\n",
    "color_maps = {'Category': category_color_map, \n",
    "              'Tissue': tissue_color_map,\n",
    "              'Array': array_color_map}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b3a7b01-7398-49e1-82fa-a9f404fa025d",
   "metadata": {},
   "source": [
    "### Sample over the years"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fcc7c64-3e17-48a1-9242-b54dd6134a0e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def sample_size_over_time(color_by = \"Category\", log_sample_size = True, model_type=\"ols\", scope = \"overall\"):\n",
    "    # Color set-up\n",
    "    if color_by in color_maps.keys():\n",
    "        color_map = color_maps[color_by]\n",
    "    else: \n",
    "        color_map = \"Virdis\"\n",
    "\n",
    "    line_color = \"grey\" if scope == \"overall\" else None\n",
    "    model_options = dict(log_y=log_sample_size) if model_type == \"ols\" else dict(frac=1)\n",
    "        \n",
    "        \n",
    "    # Plot\n",
    "    fig = px.scatter(data, x=\"Date\", y=\"Sample size\", log_y=log_sample_size, \n",
    "                     color=color_by, color_discrete_map = color_map,\n",
    "                     hover_name=data.Title.apply( lambda t: \"<br>\".join(textwrap.wrap(t, width=80))), hover_data=\"Phenotype\",\n",
    "                     trendline=model_type, trendline_scope=scope, trendline_color_override=line_color, \n",
    "                     trendline_options=model_options,\n",
    "                     title=\"Sample size over time\")\n",
    "    # Make it pretty \n",
    "    fig.update_traces(marker=dict(size=10, opacity = .5))\n",
    "    axes_style = dict(mirror=True, ticks='outside', showline=True, linecolor='black', gridcolor='lightgrey')\n",
    "    ylabel_note = \" (log scale)\" if log_sample_size else \"\"\n",
    "    fig.update_yaxes(title_text='<b>Sample size</b>'+ylabel_note, **axes_style)\n",
    "    fig.update_xaxes(title_text='Publication date', **axes_style)\n",
    "    fig.update_layout(plot_bgcolor='whitesmoke', width=1300, height=400, margin=dict(l=10, r=10, t=25, b=10))\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "sample_size_over_time()\n",
    "sample_size_over_time(log_sample_size = False)\n",
    "sample_size_over_time(model_type=\"lowess\", log_sample_size = False)\n",
    "sample_size_over_time(model_type=\"ols\", log_sample_size = False, scope=\"trace\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1148cd0b-977c-45ec-aa8f-72c9a601d35b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def category_over_years(color_by=\"Category\", percent = True, width=1300, height=400): \n",
    "    # Color set-up\n",
    "    if color_by in color_maps.keys():\n",
    "        color_map = color_maps[color_by]\n",
    "    else: \n",
    "        color_map = \"Virdis\"\n",
    "\n",
    "    norm = \"percent\" if percent else None\n",
    "    \n",
    "    fig = px.histogram(data, x=\"Year\", color=color_by, barnorm=norm,\n",
    "                   color_discrete_map = color_map,\n",
    "                   title=f\"{color_by} over time\", width=width, height=height)\n",
    "    fig.show()\n",
    "\n",
    "category_over_years()\n",
    "category_over_years(\"Tissue\", height=450)\n",
    "category_over_years(\"Array\")\n",
    "\n",
    "category_over_years(percent=False)\n",
    "category_over_years(\"Tissue\", height=450, percent=False)\n",
    "category_over_years(\"Array\", percent=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df46f0ae-b148-406a-84a5-9578cfd70d6c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig = px.histogram(data, x=\"Category\", color=\"Tissue\",\n",
    "                   color_discrete_map = tissue_color_map,\n",
    "                   category_orders = {\"Category\": list(data[\"Category\"].value_counts().index)},\n",
    "                   title=\"Category counts by tissue\", width=1400, height=450)\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d422de71-a97c-42ad-80b6-9d97f34c72e0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "counts = data[[\"Category\",\"Phenotype\"]].value_counts().reset_index()\n",
    "fig1 = px.sunburst(counts, \n",
    "                  path=['Category','Phenotype'], values='count',\n",
    "                  color = \"Category\", color_discrete_map = category_color_map, \n",
    "                  width=1200, height=750,\n",
    "                  title=\"Number of MPSs per phenotype\")\n",
    "\n",
    "fig2 = px.sunburst(data, \n",
    "                  path=['Category','Phenotype'], values='Sample size',\n",
    "                  color = \"Category\", color_discrete_map = category_color_map, \n",
    "                  width=1200, height=750,\n",
    "                  title=\"Combined sample size per phenotype\")\n",
    "fig1.show()\n",
    "fig2.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "623d9b0d-0398-4e9d-92c8-21faa0c75809",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def n_scores_by_pub(color_by=\"Category\"):\n",
    "    # Color set-up\n",
    "    if color_by in color_maps.keys():\n",
    "        color_map = color_maps[color_by]\n",
    "    else: \n",
    "        color_map = \"Virdis\"\n",
    "    \n",
    "    fig = px.histogram(data,\n",
    "                       x=\"Title\", color=color_by, color_discrete_map = color_map, \n",
    "                       nbins=186, # Number of papers\n",
    "                       # hover_name=data.Title.apply( lambda t: \"<br>\".join(textwrap.wrap(t, width=80))), \n",
    "                       # hover_data=[\"Sample size\",\"Journal\"],\n",
    "                       # category_orders = {\"Category\": list(data[\"Category\"].value_counts().index)},\n",
    "                       title=\"Number of MPSs by publication\", width=1400, height=500).update_xaxes(categoryorder=\"total ascending\")\n",
    "    # Make it pretty \n",
    "    axes_style = dict(mirror=True, showline=True, linecolor='black')\n",
    "    fig.update_yaxes(title_text='<b>Number of MPSs</b>', gridcolor='lightgrey', ticks='outside', **axes_style, \n",
    "                     tickmode = 'array',\n",
    "                     tickvals = [1, 5, 10, 20, 30, 40, 50, 70, 100])\n",
    "    fig.update_xaxes(title_text='<b>Publication</b>', **axes_style, \n",
    "                     showticklabels=False)\n",
    "    fig.update_layout(plot_bgcolor='whitesmoke', margin=dict(l=10, r=10, t=25, b=10))\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "\n",
    "n_scores_by_pub()\n",
    "n_scores_by_pub(\"Tissue\")\n",
    "n_scores_by_pub(\"Array\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a6acf32-b383-4837-975b-edc59c59c003",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create fictional age data\n",
    "age_encoding = {\n",
    "    \"Birth\": [0, 0, 0],\n",
    "    \"Very early childhood\": [1, 0, 2],\n",
    "    \"Early childhood\": [3, 2, 4],\n",
    "    \"Mid childhood\": [5, 4, 7],\n",
    "    \"Late childhood\": [9, 8, 11],\n",
    "    \"Childhood\": [5, 1, 11],\n",
    "    \"Childhood and adolescence\": [13, 10, 16],\n",
    "    \"Adolescence\": [14, 12, 18],\n",
    "    \"Not reported\": [20, 20, 20]\n",
    "}\n",
    "\n",
    "data[[\"Age_median\",\"Age_min\",\"Age_max\"]] = [age_encoding[l] for l in data[\"Developmental period\"]]\n",
    "\n",
    "# Plot\n",
    "fig = px.scatter(data, x=\"Age_median\", y=\"Tissue\", size=data[\"Sample size\"].fillna(0), size_max = 70,\n",
    "                 error_x=data[\"Age_max\"] - data[\"Age_median\"], error_x_minus=data[\"Age_median\"] - data[\"Age_min\"],\n",
    "                 hover_name=data.Title.apply( lambda t: \"<br>\".join(textwrap.wrap(t, width=80))),\n",
    "                 color=\"Category\", color_discrete_map = category_color_map)\n",
    "\n",
    "# Make it pretty \n",
    "fig.update_traces(marker=dict(opacity = .5))\n",
    "axes_style = dict(mirror=True, ticks='outside', showline=True, linecolor='black', gridcolor='lightgrey')\n",
    "fig.update_yaxes(title_text='<b>Tissue</b>', **axes_style)\n",
    "fig.update_xaxes(title_text='<b>Age of the sample</b> (years) {median, range}', **axes_style, \n",
    "                 tickmode = 'array',\n",
    "                 tickvals = [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20],\n",
    "                 ticktext = [\"Birth\", 2, 4, 6, 8, 10, 12, 14, 16, 18, \"Not reported\"])\n",
    "fig.update_layout(plot_bgcolor='whitesmoke', width=1400, height=600, margin=dict(l=10, r=10, t=25, b=10))\n",
    "\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43f7fd81-0357-4375-8904-57caa46215a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "fig = px.histogram(data, x=\"Number of CpGs\", color=\"Category\", # log_y=True, # barnorm=\"percent\",\n",
    "                   # animation_frame=\"Number of CpGs bins\", # animation_group=\"Category\",\n",
    "                   color_discrete_map = category_color_map, nbins=200,\n",
    "                   category_orders = {\"Category\": list(data[\"Category\"].value_counts().index)},\n",
    "                   title=\"Number of CpGs by category\", width=1400, height=700)\n",
    "\n",
    "# Make it pretty \n",
    "axes_style = dict(mirror=True, ticks='outside', showline=True, linecolor='black', gridcolor='lightgrey')\n",
    "fig.update_yaxes(title_text='<b>Count</b>', **axes_style)\n",
    "fig.update_xaxes(title_text='<b>Number of CpGs</b>', **axes_style)\n",
    "                 # tickmode = 'array',\n",
    "                 # tickvals = [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000],\n",
    "                 # ticktext = [\"1\", \"1k\", \"2k\", \"3k\", \"4k\", \"5k\", \"6k\", \"7k\", \"8k\", \"9k\", \">10k\"])\n",
    "\n",
    "# Add range slider\n",
    "fig.update_layout(\n",
    "    plot_bgcolor='whitesmoke',\n",
    "    xaxis=dict(\n",
    "        # autorange=False,\n",
    "        # range = [0, 1000], \n",
    "        rangeslider=dict(visible=True, \n",
    "                         autorange=False,\n",
    "                         range=[0,1000]),\n",
    "        type=\"linear\"\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acf933dd-f1b8-466d-8bc9-ab4a3ba2b06e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b32f41-df18-4f9c-bca8-53597d49d9a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean up: TMP\n",
    "data[\"Number of CpGs\"] = pd.to_numeric(data[\"Number of CpGs\"], errors='coerce')\n",
    "data[\"Number of CpGs crop\"] = [10_000 if i > 10_000 else i for i in data[\"Number of CpGs\"]]\n",
    "data[\"Number of CpGs bins\"] = pd.cut(data[\"Number of CpGs\"],\n",
    "                                     bins=[0, 1_000, 5_000, 10_000, 160_000], include_lowest=True,\n",
    "                                     labels=['< 1k', '1 to 5k', '5 to 10k', \"> 10k\"])\n",
    "\n",
    "# Plot\n",
    "fig = px.histogram(data, x=\"Number of CpGs\", color=\"Category\", # log_y=True, # barnorm=\"percent\",\n",
    "                   animation_frame=\"Number of CpGs bins\", # animation_group=\"Category\",\n",
    "                   color_discrete_map = category_color_map, nbins=150,\n",
    "                   category_orders = {\"Category\": list(data[\"Category\"].value_counts().index)},\n",
    "                   title=\"Number of CpGs by category\", width=1400, height=700)\n",
    "\n",
    "# Make it pretty \n",
    "axes_style = dict(mirror=True, ticks='outside', showline=True, linecolor='black', gridcolor='lightgrey')\n",
    "fig.update_yaxes(title_text='<b>Count</b>', **axes_style)\n",
    "fig.update_xaxes(title_text='<b>Number of CpGs</b>', **axes_style)\n",
    "                 # tickmode = 'array',\n",
    "                 # tickvals = [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000],\n",
    "                 # ticktext = [\"1\", \"1k\", \"2k\", \"3k\", \"4k\", \"5k\", \"6k\", \"7k\", \"8k\", \"9k\", \">10k\"])\n",
    "fig.update_layout(plot_bgcolor='whitesmoke')\n",
    "\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c3c322-0674-48e8-af33-17bc436a424f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.scatter(data, x=\"Category\", y=\"Number of CpGs\", animation_frame=\"Year\", animation_group=\"Title\",\n",
    "                 color=\"Tissue\", color_discrete_map = tissue_color_map,\n",
    "                 size=data[\"Sample size\"].fillna(0), \n",
    "                 hover_name=\"Title\", facet_col=\"Array\",\n",
    "                 width=1400, height=650,\n",
    "                 title=\"Bit of everything\"\n",
    "                 )\n",
    "           # log_x=True, size_max=45, range_x=[100,100000], range_y=[25,90])\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c49f8981-a2ec-4a31-94b7-178fcfea0161",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.scatter(data, x=\"Number of CpGs\", y=\"Effect size phenotype\", size=data[\"Sample size\"].fillna(0), \n",
    "                 color=\"Tissue\", hover_name=\"Title\")\n",
    "           # log_x=True, size_max=45, range_x=[100,100000], range_y=[25,90])\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4680b1b-2bf3-48a3-93f3-9596c146b1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.bar_polar(pd.DataFrame(data[[\"Category\", \"Tissue\"]].value_counts()).reset_index(), \n",
    "                   r=\"count\", theta=\"Category\", color=\"Tissue\", width=900, height=900) #, template=\"plotly_dark\", #  line_close=True,\n",
    "            # color_discrete_sequence=px.colors.sequential.Plasma_r)\n",
    "fig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1387336a-aae0-4a33-9471-c0ac751214b5",
   "metadata": {},
   "source": [
    "# Abstract analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aac09af-c87c-467e-899f-ee876aaece22",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a2f85cd-7977-4302-8d8b-6f8916d7eacd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# text = bib.Abstract\n",
    "\n",
    "text = \" \".join(abstract for abstract in bib.Abstract.fillna(\"\"))\n",
    "print (\"There are {} words in the combination of all review.\".format(len(text)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b204be44-9485-4aee-923d-befe2f37151f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create stopword list:\n",
    "stopwords = set(STOPWORDS)\n",
    "stopwords.update([\"CpG\",\"CpGs\",\"DNA\",\"DNAm\",\"methylation\",\"episignature\",\"episignatures\",\"epigenetic\",\"gene\",\"genes\",\"genetic\",\"variant\",\"variants\",\n",
    "                  \"score\",\"scores\",\"result\",\"results\",\"analysis\",\"study\",\"studies\",\"cohort\",\"sample\",\"n\",\"p\",\"CI\",\"data\",\"significant\",\n",
    "                  \"background\",\"conclusion\",\"copyright\",\"model\",\"finding\",\"method\",\"methods\",\n",
    "                  \"associated\",\"association\",\"associations\",\"effect\",\"effects\",\"related\",\"identify\",\"identified\",\"provide\",\"assessed\",\"using\",\"used\",\"based\",\n",
    "                  \"specific\",\"distinct\",\"one\",\"two\",\"three\", \"may\",\"including\",\"within\",\n",
    "                  \"patient\",\"patients\",\"individual\",\"individuals\",\"subject\", \"case\",\"cases\"])\n",
    "\n",
    "# Generate a word cloud image\n",
    "wordcloud = WordCloud(stopwords=stopwords, background_color=\"white\").generate(text)\n",
    "\n",
    "# Display the generated image\n",
    "fig, ax = plt.subplots(figsize=(5, 5), facecolor='lightskyblue', layout='constrained')\n",
    "# the matplotlib way:\n",
    "ax.imshow(wordcloud, interpolation='bilinear')\n",
    "ax.axis(\"off\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1df9201-42b3-450a-b178-4d9346b1b40e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def wordcould_by_category(category, extra_stopwords=[]):\n",
    "    subcat = bib.loc[bib.Identifier.isin([int(i) for i in pd.unique(data.loc[data.Category==category, \"Identifier\"])])]\n",
    "    \n",
    "    text = \" \".join(abstract for abstract in subcat.Abstract.fillna(\"\"))\n",
    "\n",
    "    # Create stopword list:\n",
    "    stopwords = set(STOPWORDS)\n",
    "    stopwords.update([\"CpG\",\"CpGs\",\"DNA\",\"DNAm\",\"methylation\",\"episignature\",\"episignatures\",\"epigenetic\",\"gene\",\"genes\",\"genetic\",\"variant\",\"variants\",\n",
    "                      \"score\",\"scores\",\"result\",\"results\",\"analysis\",\"study\",\"studies\",\"cohort\",\"sample\",\"n\",\"p\",\"CI\",\"data\",\"significant\",\n",
    "                      \"background\",\"conclusion\",\"copyright\",\"model\",\"finding\",\"method\",\"methods\",\n",
    "                      \"associated\",\"association\",\"associations\",\"effect\",\"effects\",\"related\",\"identify\",\"identified\",\"provide\",\"assessed\",\"using\",\"used\",\"based\",\n",
    "                      \"specific\",\"distinct\",\"one\",\"two\",\"three\", \"may\",\"including\",\"within\",\n",
    "                      \"patient\",\"patients\",\"individual\",\"individuals\",\"subject\", \"case\",\"cases\"]+extra_stopwords)\n",
    "    \n",
    "    # Generate a word cloud image\n",
    "    wordcloud = WordCloud(stopwords=stopwords, background_color=\"white\").generate(text)\n",
    "    \n",
    "    # Display the generated image\n",
    "    fig, ax = plt.subplots(figsize=(10, 5), facecolor=category_color_map[category], layout='constrained')\n",
    "    # the matplotlib way:\n",
    "    ax.imshow(wordcloud, interpolation='bilinear')\n",
    "    ax.set_title(category, fontsize=18, fontweight=\"bold\")\n",
    "    ax.text(1,1.1, \"Based on {} words from {} abstracts.\".format(len(text.split(\" \")), subcat.shape[0]), fontsize=12,\n",
    "            ha=\"right\", transform=ax.transAxes)\n",
    "    ax.axis(\"off\");\n",
    "\n",
    "for c in data.Category.unique(): wordcould_by_category(c)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43976665-de50-4b4d-846a-42a441d5f28f",
   "metadata": {},
   "source": [
    "# LLM playground"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27de6bb6-a092-4759-b9d3-17130cc95f9f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from llama_cpp import Llama\n",
    "# model source: https://huggingface.co/TheBloke/orca_mini_v3_7B-GGUF \n",
    "# Bits: 4\tSize: 4.08 GB\tMax RAM required: 6.58 GB\n",
    "LLM = Llama(model_path=\"../orca_mini_v3_7b.Q4_K_M.gguf\", n_ctx=2048) # medium, balanced quality - recommended # max_new_tokens = 4096, max_tokens = 2000,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b5f3f36-ed34-4e8c-afd4-1ab56dff2ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a text prompt\n",
    "prompt = f'Based on the following abstract, what was the sample size of the study and what statistical or machine-learning and what method was used: \\n\\n\"{bib.Abstract[10]}\"'\n",
    "\n",
    "# generate a response (takes several seconds)\n",
    "output = LLM(prompt, max_tokens=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5682674d-7b65-4424-adc8-47f07a68007d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# display the response\n",
    "print(prompt, \"\\n\\n----\", output[\"choices\"][0][\"text\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "demetra_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
