{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "efd73777-783a-4f41-a3a4-bd29b7460112",
   "metadata": {},
   "source": [
    "# DeMetRA - literature review\n",
    "\n",
    "Read and clean dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ff8019b0-3a9d-4e86-b6fc-87b3e65eeb5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from dateutil.parser import parse as parse_date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4d4a974-cbdb-43da-af7b-582845b4c988",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Serena/Desktop/DEMETRA/DeMetRA-review/demetra_venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n",
      "/Users/Serena/Desktop/DEMETRA/DeMetRA-review/demetra_venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "assets_directory = '../assets/'\n",
    "\n",
    "lit, base_ss, base_va = pd.read_excel(f'{assets_directory}MPS_review_systematic_2025-02-14.xlsx', sheet_name=[0, 1, 2]).values()\n",
    "\n",
    "# lit = lit.loc[lit.Include == 'Yes'].drop('Include', axis=1) # All included\n",
    "\n",
    "# print(lit.shape, '\\n', list(lit.columns))\n",
    "# print(base_ss.shape, '\\n', list(base_ss.columns))\n",
    "# print(base_va.shape, '\\n', list(base_va.columns))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c14dc6c7-988f-4e57-bcc2-2745f57d26c6",
   "metadata": {},
   "source": [
    "### Data inspection and cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e35eac17-3145-4027-8ad6-a56e07c9690c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tissue\n",
      "Peripheral blood          497\n",
      "Whole blood               153\n",
      "Saliva                    120\n",
      "Cord blood                 48\n",
      "Dried bloodspot            26\n",
      "Blood-clots                26\n",
      "Buccal cells               21\n",
      "Placenta                   15\n",
      "Tumour cells                7\n",
      "Nasal epithelial cells      7\n",
      "Multiple                    4\n",
      "Leukocytes                  4\n",
      "Not reported                1\n",
      "Cervical cells              1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Multiple_tissue\n",
      "NaN                            926\n",
      "Whole blood, HPCs                2\n",
      "Cord blood, Dried bloodspot      1\n",
      "Cord blood, Whole blood          1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Array\n",
      "Multiple    436\n",
      "EPICv1      289\n",
      "450K        191\n",
      "WGBS         14\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Multiple_array\n",
      "NaN                                    494\n",
      "450K, EPICv1                           426\n",
      "450K, GMEL (~3000 CpGs from EPICv1)      6\n",
      "450K, EPICv2                             2\n",
      "450K, EPICv3                             1\n",
      "450K, EPICv4                             1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Ancestry\n",
      "European        497\n",
      "Not reported    227\n",
      "White           145\n",
      "Multiple         48\n",
      "Mixed             8\n",
      "African           2\n",
      "Hispanic          2\n",
      "Australian        1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Multiple_ancestry\n",
      "NaN                                                       874\n",
      "White, African-American, Indian-American                   14\n",
      "Mixed, African, European                                    8\n",
      "European, Hispanic/Latinx                                   6\n",
      "Mixed, African                                              6\n",
      "White, African-American, Hispanic                           4\n",
      "White, African-American, Hispanic/Latinx, Asian, Other      2\n",
      "European, African-American, Hispanic/Latinx                 2\n",
      "White, Hispanic/Latinx, Other                               2\n",
      "European, North-American                                    2\n",
      "European, African-American                                  2\n",
      "White, African-American, Hispanic/Latinx, Asian             1\n",
      "White, African-American, Asian                              1\n",
      "European, Hispanic/Latinx , Mixed, African, American        1\n",
      "European, African, Hispanic                                 1\n",
      "White, other                                                1\n",
      "White, African-American, Hispanic/Latinx                    1\n",
      "European, African, other                                    1\n",
      "European, Indian Asian                                      1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Developmental_period\n",
      "Mid childhood                       321\n",
      "Birth                               174\n",
      "Childhood and adolescence           134\n",
      "Adults                              117\n",
      "Not reported                         94\n",
      "Late childhood                       45\n",
      "Adolescence                          29\n",
      "Very early childhood                 11\n",
      "Early childhood                       3\n",
      "Birth, Childhood and adolescence      1\n",
      "Childhood                             1\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Inspect values \n",
    "def checklvl(var):\n",
    "    allobs = pd.concat([lit[var], base_ss[var], base_va[var]])\n",
    "    print(allobs.value_counts(dropna=False))\n",
    "\n",
    "for v in  ['Tissue', 'Array','Ancestry','Developmental_period']:\n",
    "    checklvl(v)\n",
    "    print('\\n')\n",
    "    try: \n",
    "        checklvl(f'Multiple_{v.lower()}')\n",
    "        print('\\n')\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "# NOTE: this is to guide manual cleaning of the excel file!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d6616b2e-205f-4f33-aa09-ea8b381fb2d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recode \"multiple\" categories\n",
    "def replace_multiples(df):\n",
    "\n",
    "    for var in ['Tissue', 'Array', 'Ancestry']:\n",
    "    \n",
    "        if var == 'Ancestry':\n",
    "            df.loc[df[var] == 'Multiple', var] = 'Mixed'\n",
    "    \n",
    "        else:\n",
    "            df.loc[df[var] == 'Multiple', var] = [f'Multiple ({values})' for values in df.loc[df[var] == 'Multiple', f'Multiple_{var.lower()}']]\n",
    "        \n",
    "        # print(df[var].value_counts(), '\\n')\n",
    "    \n",
    "    return df\n",
    "\n",
    "lit = replace_multiples(lit)\n",
    "base_ss = replace_multiples(base_ss)\n",
    "base_va = replace_multiples(base_va)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fca1748c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Type</th>\n",
       "      <th>Author</th>\n",
       "      <th>Year</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Include</th>\n",
       "      <th>pdf</th>\n",
       "      <th>Sample_size_total</th>\n",
       "      <th>...</th>\n",
       "      <th>Including_CpGs_4</th>\n",
       "      <th>Including_CpGs_5</th>\n",
       "      <th>Number of CpGs</th>\n",
       "      <th>Determining_weights_1</th>\n",
       "      <th>Train_test</th>\n",
       "      <th>Independent_validation</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>Missing_value_note</th>\n",
       "      <th>Reflect_phenotype</th>\n",
       "      <th>Link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Abrishamcar (2022) - DNA methylation as a pote...</td>\n",
       "      <td>262</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Discovery EWAS | linear regression</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | Explained variance | R2 = 0.23</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3_x...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Abrishamcar (2022) - DNA methylation as a pote...</td>\n",
       "      <td>262</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Discovery EWAS | linear regression</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Reported | Explained variance | R = 0.0006</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3_x...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>R. Al-Jawahiri; A. Foroutan; J. Kerkhof; H. Mc...</td>\n",
       "      <td>2022</td>\n",
       "      <td>SOX11 variants cause a neurodevelopmental diso...</td>\n",
       "      <td>Gen Med</td>\n",
       "      <td>10.1016/j.gim.2022.02.013</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Al-Jawahiri (2022) - SOX11 variants cause a ne...</td>\n",
       "      <td>126</td>\n",
       "      <td>...</td>\n",
       "      <td>Pruning | Pairwise correlation | 0.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>224</td>\n",
       "      <td>Machine learning | Support vector machine</td>\n",
       "      <td>Split 25/75%, 10-fold cross-validation</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No actual number reported</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Identifier             Type  \\\n",
       "0           1  Journal Article   \n",
       "1           1  Journal Article   \n",
       "2           3  Journal Article   \n",
       "\n",
       "                                              Author  Year  \\\n",
       "0  S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...  2022   \n",
       "1  S. Abrishamcar; J. Chen; D. Feil; A. Kilanowsk...  2022   \n",
       "2  R. Al-Jawahiri; A. Foroutan; J. Kerkhof; H. Mc...  2022   \n",
       "\n",
       "                                               Title            Journal  \\\n",
       "0  DNA methylation as a potential mediator of the...  Transl Psychiatry   \n",
       "1  DNA methylation as a potential mediator of the...  Transl Psychiatry   \n",
       "2  SOX11 variants cause a neurodevelopmental diso...            Gen Med   \n",
       "\n",
       "                          DOI Include  \\\n",
       "0  10.1038/s41398-022-02195-3     Yes   \n",
       "1  10.1038/s41398-022-02195-3     Yes   \n",
       "2   10.1016/j.gim.2022.02.013     Yes   \n",
       "\n",
       "                                                 pdf Sample_size_total  ...  \\\n",
       "0  Abrishamcar (2022) - DNA methylation as a pote...               262  ...   \n",
       "1  Abrishamcar (2022) - DNA methylation as a pote...               262  ...   \n",
       "2  Al-Jawahiri (2022) - SOX11 variants cause a ne...               126  ...   \n",
       "\n",
       "                       Including_CpGs_4 Including_CpGs_5 Number of CpGs  \\\n",
       "0                                   NaN              NaN   Not reported   \n",
       "1                                   NaN              NaN   Not reported   \n",
       "2  Pruning | Pairwise correlation | 0.6              NaN            224   \n",
       "\n",
       "                       Determining_weights_1  \\\n",
       "0         Discovery EWAS | linear regression   \n",
       "1         Discovery EWAS | linear regression   \n",
       "2  Machine learning | Support vector machine   \n",
       "\n",
       "                               Train_test Independent_validation Comparison  \\\n",
       "0                                      No                     No         No   \n",
       "1                                      No                     No         No   \n",
       "2  Split 25/75%, 10-fold cross-validation                     No         No   \n",
       "\n",
       "  Missing_value_note                           Reflect_phenotype  \\\n",
       "0       Not reported   Reported | Explained variance | R2 = 0.23   \n",
       "1       Not reported  Reported | Explained variance | R = 0.0006   \n",
       "2                 No                   No actual number reported   \n",
       "\n",
       "                                                Link  \n",
       "0  http://dx.doi.org/10.1038/s41398-022-02195-3_x...  \n",
       "1  http://dx.doi.org/10.1038/s41398-022-02195-3_x...  \n",
       "2  https://www.embase.com/search/results?subactio...  \n",
       "\n",
       "[3 rows x 40 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lit.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f61a6d7",
   "metadata": {},
   "source": [
    "### Parse bibliography for publication dates, abstracts and keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b66abae5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Author_list</th>\n",
       "      <th>Year</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Date</th>\n",
       "      <th>DOI</th>\n",
       "      <th>URL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[Abrishamcar, S., Chen, J., Feil, D., Kilanows...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prenatal tobacco exposure (PTE) and prenatal a...</td>\n",
       "      <td>09 30</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[Ács, O., Péterfia, B., Hollósi, P., Luczay, A...</td>\n",
       "      <td>2017</td>\n",
       "      <td>Methylation Status of CYP27B1 and IGF2 Correla...</td>\n",
       "      <td>Obes Facts</td>\n",
       "      <td>[25 hydroxyvitamin D, bisulfite, cytochrome P4...</td>\n",
       "      <td>Objective: Worldwide increasing childhood obes...</td>\n",
       "      <td>2017-08-16</td>\n",
       "      <td>10.1159/000477462</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[Adcock, R., Nedjai, B., Lorincz, A. T., Scibi...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation testing with S5 for triage of ...</td>\n",
       "      <td>Int J Cancer</td>\n",
       "      <td>[adolescent, adult, aged, article, biopsy, can...</td>\n",
       "      <td>Methylation of host and viral genes is promisi...</td>\n",
       "      <td>2022-06-27</td>\n",
       "      <td>10.1002/ijc.34050</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[Al-Jawahiri, R., Foroutan, A., Kerkhof, J., M...</td>\n",
       "      <td>2022</td>\n",
       "      <td>SOX11 variants cause a neurodevelopmental diso...</td>\n",
       "      <td>Gen Med</td>\n",
       "      <td>[nucleic acid genotyping array kit, growth dif...</td>\n",
       "      <td>Purpose: This study aimed to undertake a multi...</td>\n",
       "      <td>2022-04-05</td>\n",
       "      <td>10.1016/j.gim.2022.02.013</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[Alfano, R., Zugna, D., Barros, H., Bustamante...</td>\n",
       "      <td>2023</td>\n",
       "      <td>Cord blood epigenome-wide meta-analysis in six...</td>\n",
       "      <td>BMC Med</td>\n",
       "      <td>[aurora C kinase, long untranslated RNA, ACTG1...</td>\n",
       "      <td>Background: Rapid postnatal growth may result ...</td>\n",
       "      <td>2023-01-17</td>\n",
       "      <td>10.1186/s12916-022-02685-7</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Author_list  Year  \\\n",
       "0  [Abrishamcar, S., Chen, J., Feil, D., Kilanows...  2022   \n",
       "1  [Ács, O., Péterfia, B., Hollósi, P., Luczay, A...  2017   \n",
       "2  [Adcock, R., Nedjai, B., Lorincz, A. T., Scibi...  2022   \n",
       "3  [Al-Jawahiri, R., Foroutan, A., Kerkhof, J., M...  2022   \n",
       "4  [Alfano, R., Zugna, D., Barros, H., Bustamante...  2023   \n",
       "\n",
       "                                               Title            Journal  \\\n",
       "0  DNA methylation as a potential mediator of the...  Transl Psychiatry   \n",
       "1  Methylation Status of CYP27B1 and IGF2 Correla...         Obes Facts   \n",
       "2  DNA methylation testing with S5 for triage of ...       Int J Cancer   \n",
       "3  SOX11 variants cause a neurodevelopmental diso...            Gen Med   \n",
       "4  Cord blood epigenome-wide meta-analysis in six...            BMC Med   \n",
       "\n",
       "                                            Keywords  \\\n",
       "0                                                NaN   \n",
       "1  [25 hydroxyvitamin D, bisulfite, cytochrome P4...   \n",
       "2  [adolescent, adult, aged, article, biopsy, can...   \n",
       "3  [nucleic acid genotyping array kit, growth dif...   \n",
       "4  [aurora C kinase, long untranslated RNA, ACTG1...   \n",
       "\n",
       "                                            Abstract        Date  \\\n",
       "0  Prenatal tobacco exposure (PTE) and prenatal a...       09 30   \n",
       "1  Objective: Worldwide increasing childhood obes...  2017-08-16   \n",
       "2  Methylation of host and viral genes is promisi...  2022-06-27   \n",
       "3  Purpose: This study aimed to undertake a multi...  2022-04-05   \n",
       "4  Background: Rapid postnatal growth may result ...  2023-01-17   \n",
       "\n",
       "                          DOI  \\\n",
       "0  10.1038/s41398-022-02195-3   \n",
       "1           10.1159/000477462   \n",
       "2           10.1002/ijc.34050   \n",
       "3   10.1016/j.gim.2022.02.013   \n",
       "4  10.1186/s12916-022-02685-7   \n",
       "\n",
       "                                                 URL  \n",
       "0       http://dx.doi.org/10.1038/s41398-022-02195-3  \n",
       "1  https://www.embase.com/search/results?subactio...  \n",
       "2  https://www.embase.com/search/results?subactio...  \n",
       "3  https://www.embase.com/search/results?subactio...  \n",
       "4  https://www.embase.com/search/results?subactio...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parsing bibliography file (this shoud be a RIS or TXT file in RIS format)\n",
    "\n",
    "ris_tags = {\n",
    "    'TY': 'Reference Type',\n",
    "    'AU': 'Author_list',\n",
    "    'PY': 'Year',\n",
    "    'TI': 'Title', # 'T1\n",
    "    'T2': 'Journal', # 'JO'\n",
    "    'J2': 'Journal', # For pre-prints \n",
    "    'AB': 'Abstract',\n",
    "    'DO': 'DOI',\n",
    "    'UR': 'URL',\n",
    "    'KW': 'Keywords',\n",
    "    'DA': 'Date'\n",
    "}\n",
    "\n",
    "def parse_ris(file_path):\n",
    "    with open(file_path, 'r', encoding='utf-8') as file:\n",
    "        references = []\n",
    "\n",
    "        current_entry = {}\n",
    "        current_tag = None\n",
    "\n",
    "        for line in file:\n",
    "            if not line.strip():  # Skip empty lines\n",
    "                continue\n",
    "            elif line.strip() == 'ER  -':  # ER marks the end of the record\n",
    "                # Append reference and reset\n",
    "                references.append(current_entry)\n",
    "                current_entry = {}\n",
    "                current_tag = None \n",
    "            else:\n",
    "                tag = line[:5]\n",
    "\n",
    "                if tag in [f'{t}  -' for t in ris_tags.keys()]:\n",
    "                    current_tag = tag\n",
    "\n",
    "                    value = line[5:].strip()\n",
    "                    key = ris_tags[tag[:2]]\n",
    "\n",
    "                    if key in current_entry:\n",
    "                        if isinstance(current_entry[key], list):\n",
    "                            current_entry[key].append(value)\n",
    "                        else:\n",
    "                            current_entry[key] = [current_entry[key], value]\n",
    "                    else:\n",
    "                        current_entry[key] = value\n",
    "\n",
    "                # Handle dates that span multiple lines\n",
    "                elif tag in [f'{y}-' for y in range(2000, 2025)]:\n",
    "                    value = line.strip()\n",
    "                    if 'Date' not in current_entry:\n",
    "                        current_entry['Date'] = value # Only keep first occurrance (i.e. publication date)\n",
    "\n",
    "                # Handle keywords that span multiple lines\n",
    "                else:\n",
    "                    if current_tag == 'KW  -':\n",
    "                        value = line.strip()\n",
    "                        if isinstance(current_entry['Keywords'], list):\n",
    "                            current_entry['Keywords'].append(value)\n",
    "                        else:\n",
    "                            current_entry['Keywords'] = [current_entry['Keywords'], value]\n",
    "                                    \n",
    "        return references\n",
    "\n",
    "\n",
    "parsed_data = parse_ris(f'{assets_directory}Bibliography_2024-09-19.txt')\n",
    "bib = pd.DataFrame(parsed_data)[['Author_list', 'Year', 'Title', 'Journal', 'Keywords', 'Abstract', \n",
    "                                 'Date', 'DOI', 'URL']]\n",
    "\n",
    "bib.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1865ebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note 9 NaN Date values will be set to 01/01 of respective year\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Author_list</th>\n",
       "      <th>Year</th>\n",
       "      <th>Title</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Abstract</th>\n",
       "      <th>Date</th>\n",
       "      <th>DOI</th>\n",
       "      <th>URL</th>\n",
       "      <th>Author</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[Abrishamcar, S., Chen, J., Feil, D., Kilanows...</td>\n",
       "      <td>2022</td>\n",
       "      <td>DNA methylation as a potential mediator of the...</td>\n",
       "      <td>Transl Psychiatry</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prenatal tobacco exposure (PTE) and prenatal a...</td>\n",
       "      <td>2022-09-30</td>\n",
       "      <td>10.1038/s41398-022-02195-3</td>\n",
       "      <td>http://dx.doi.org/10.1038/s41398-022-02195-3</td>\n",
       "      <td>Abrishamcar et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[Al-Jawahiri, R., Foroutan, A., Kerkhof, J., M...</td>\n",
       "      <td>2022</td>\n",
       "      <td>SOX11 variants cause a neurodevelopmental diso...</td>\n",
       "      <td>Gen Med</td>\n",
       "      <td>[nucleic acid genotyping array kit, growth dif...</td>\n",
       "      <td>Purpose: This study aimed to undertake a multi...</td>\n",
       "      <td>2022-04-05</td>\n",
       "      <td>10.1016/j.gim.2022.02.013</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Al-Jawahiri et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[Alfano, R., Zugna, D., Barros, H., Bustamante...</td>\n",
       "      <td>2023</td>\n",
       "      <td>Cord blood epigenome-wide meta-analysis in six...</td>\n",
       "      <td>BMC Med</td>\n",
       "      <td>[aurora C kinase, long untranslated RNA, ACTG1...</td>\n",
       "      <td>Background: Rapid postnatal growth may result ...</td>\n",
       "      <td>2023-01-17</td>\n",
       "      <td>10.1186/s12916-022-02685-7</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Alfano et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[Aref-Eshghi, E., Bend, E. G., Colaiacovo, S.,...</td>\n",
       "      <td>2019</td>\n",
       "      <td>Diagnostic Utility of Genome-wide DNA Methylat...</td>\n",
       "      <td>Am J Hum Genet</td>\n",
       "      <td>[genomic DNA, adult, article, case report, chi...</td>\n",
       "      <td>Conventional genetic testing of individuals wi...</td>\n",
       "      <td>2019-04-05</td>\n",
       "      <td>10.1016/j.ajhg.2019.03.008</td>\n",
       "      <td>https://www.embase.com/search/results?subactio...</td>\n",
       "      <td>Aref-Eshghi et al.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[Aref-Eshghi, E., Kerkhof, J., Pedro, V. P.]</td>\n",
       "      <td>2020</td>\n",
       "      <td>Evaluation of DNA methylation episignatures fo...</td>\n",
       "      <td>The American Journal of …</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-01-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aref-Eshghi et al.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Author_list  Year  \\\n",
       "0  [Abrishamcar, S., Chen, J., Feil, D., Kilanows...  2022   \n",
       "1  [Al-Jawahiri, R., Foroutan, A., Kerkhof, J., M...  2022   \n",
       "2  [Alfano, R., Zugna, D., Barros, H., Bustamante...  2023   \n",
       "3  [Aref-Eshghi, E., Bend, E. G., Colaiacovo, S.,...  2019   \n",
       "4       [Aref-Eshghi, E., Kerkhof, J., Pedro, V. P.]  2020   \n",
       "\n",
       "                                               Title  \\\n",
       "0  DNA methylation as a potential mediator of the...   \n",
       "1  SOX11 variants cause a neurodevelopmental diso...   \n",
       "2  Cord blood epigenome-wide meta-analysis in six...   \n",
       "3  Diagnostic Utility of Genome-wide DNA Methylat...   \n",
       "4  Evaluation of DNA methylation episignatures fo...   \n",
       "\n",
       "                     Journal  \\\n",
       "0          Transl Psychiatry   \n",
       "1                    Gen Med   \n",
       "2                    BMC Med   \n",
       "3             Am J Hum Genet   \n",
       "4  The American Journal of …   \n",
       "\n",
       "                                            Keywords  \\\n",
       "0                                                NaN   \n",
       "1  [nucleic acid genotyping array kit, growth dif...   \n",
       "2  [aurora C kinase, long untranslated RNA, ACTG1...   \n",
       "3  [genomic DNA, adult, article, case report, chi...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                            Abstract        Date  \\\n",
       "0  Prenatal tobacco exposure (PTE) and prenatal a...  2022-09-30   \n",
       "1  Purpose: This study aimed to undertake a multi...  2022-04-05   \n",
       "2  Background: Rapid postnatal growth may result ...  2023-01-17   \n",
       "3  Conventional genetic testing of individuals wi...  2019-04-05   \n",
       "4                                                NaN  2020-01-01   \n",
       "\n",
       "                          DOI  \\\n",
       "0  10.1038/s41398-022-02195-3   \n",
       "1   10.1016/j.gim.2022.02.013   \n",
       "2  10.1186/s12916-022-02685-7   \n",
       "3  10.1016/j.ajhg.2019.03.008   \n",
       "4                         NaN   \n",
       "\n",
       "                                                 URL              Author  \n",
       "0       http://dx.doi.org/10.1038/s41398-022-02195-3  Abrishamcar et al.  \n",
       "1  https://www.embase.com/search/results?subactio...  Al-Jawahiri et al.  \n",
       "2  https://www.embase.com/search/results?subactio...       Alfano et al.  \n",
       "3  https://www.embase.com/search/results?subactio...  Aref-Eshghi et al.  \n",
       "4                                                NaN  Aref-Eshghi et al.  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean bibliography file\n",
    "\n",
    "# Only selected papers ----------------------------------------------------------------\n",
    "bib_incl = bib.loc[bib.Title.isin(lit.Title.unique()), ].reset_index(drop=True)\n",
    "# bib_incl.shape\n",
    "\n",
    "# Correct dates ----------------------------------------------------------------------- \n",
    "print('Note', bib_incl.Date.isna().sum(), 'NaN Date values will be set to 01/01 of respective year')\n",
    "\n",
    "date_tmp = pd.Series([' '.join([d, y]) if y not in d else d for d, y in \n",
    "                   zip(bib_incl.Date.map(str), bib_incl.Year.map(str))])\n",
    "\n",
    "bib_incl.loc[:, 'Date'] = date_tmp.apply(lambda date: \n",
    "                                         parse_date(date).strftime('%Y-%m-%d') if 'nan' not in date else \n",
    "                                         parse_date('01 01' + date[3:]).strftime('%Y-%m-%d'))\n",
    "\n",
    "# Get Short Author titles ------------------------------------------------------------\n",
    "bib_incl['Author'] = [f'{fa[0].split(\",\")[0]} et al.' if len(fa) > 1 else f'{fa[0].split(\",\")[0]}' \n",
    "                            for fa in bib_incl['Author_list']]\n",
    "\n",
    "\n",
    "# Fix missing Journal names ----------------------------------------------------------\n",
    "bib_incl.loc[bib_incl.DOI.str.contains('10.3390/toxics9100262', na=False), \n",
    "             'Journal'] = 'Toxics'\n",
    "bib_incl.loc[bib_incl.DOI.str.contains('10.1007/s00787-024-02390-1', na=False), \n",
    "             'Journal'] = 'Eur Child Adolesc Psych'\n",
    "bib_incl.loc[bib_incl.DOI.str.contains('10.3390/ijms22031111|10.3390/ijms22168611', na=False), \n",
    "             'Journal'] = 'Int J Mol Sci'\n",
    "\n",
    "bib_incl.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bcc2eb79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(760, 40)\n",
      "(760, 47)\n",
      "Index(['Identifier', 'Type', 'Author_dirty', 'Year', 'Title', 'Journal_dirty',\n",
      "       'DOI', 'Include', 'pdf', 'Sample_size_total', 'Sample_size_control',\n",
      "       'Sample_size_case', 'Sample_type', 'Category', 'Phenotype',\n",
      "       'What_is_available', 'Identifier_base', 'Multiple_identifier_base',\n",
      "       'Tissue', 'Multiple_tissue', 'Array', 'Multiple_array', 'Ancestry',\n",
      "       'Multiple_ancestry', 'Developmental_period', 'Covariates',\n",
      "       'Sample_overlap_target_base', 'Including_CpGs_1', 'Including_CpGs_2',\n",
      "       'Including_CpGs_3', 'Including_CpGs_4', 'Including_CpGs_5',\n",
      "       'Number of CpGs', 'Determining_weights_1', 'Train_test',\n",
      "       'Independent_validation', 'Comparison', 'Missing_value_note',\n",
      "       'Reflect_phenotype', 'Link', 'Author_list', 'Journal', 'Keywords',\n",
      "       'Abstract', 'Date', 'URL', 'Author'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(lit.shape)\n",
    "\n",
    "lit_main = lit.rename(columns={'Author': 'Author_dirty',\n",
    "                               'Journal': 'Journal_dirty'}).merge(bib_incl,\n",
    "                     on='Title', how='left', suffixes=['','_BIB'])\n",
    "\n",
    "# CHECKUPS \n",
    "# lit_main[['Author','Author_list','Author_dirty']]\n",
    "# np.where(lit_main['Year'].map(int) != lit_main['Year_BIB'].map(int))\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# lit_main.loc[lit_main['DOI'].map(str) != lit_main['DOI_BIB'].map(str), ['Title','DOI','DOI_BIB']]\n",
    "\n",
    "# Clean DOI values -------------------------------------------------------------------\n",
    "lit_main['DOI_BIB'] = lit_main['DOI_BIB'].fillna(lit_main['DOI'])\n",
    "lit_main['DOI'] = lit_main['DOI_BIB']\n",
    "lit_main.drop(['DOI_BIB','Year_BIB'], axis=1, inplace=True)\n",
    "\n",
    "print(lit_main.shape)\n",
    "print(lit_main.columns)\n",
    "\n",
    "# Save cleaned data ------------------------------------------------------------------\n",
    "lit_main.to_csv(f'{assets_directory}MPS_literature_cleaned.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "459d3e58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Array\n",
       "Multiple (450K, EPICv1)                           393\n",
       "EPICv1                                            283\n",
       "450K                                               63\n",
       "WGBS                                               14\n",
       "Multiple (450K, GMEL (~3000 CpGs from EPICv1))      6\n",
       "Multiple (450K, EPICv2)                             1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lit_main.Array.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4456be9e",
   "metadata": {},
   "source": [
    "### Merge base and target info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f660d332-52e8-4c2f-bb20-7f24760d2136",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(170, 9) \n",
      " ['Identifier', 'Category', 'Phenotype', 'Tissue', 'Array', 'Ancestry', 'Developmental_period', 'Covariates', 'MPS_id']\n",
      "(283, 13) \n",
      " ['Identifier', 'Title', 'Year', 'What_is_available', 'Sample_overlap_target_base', 'Category', 'Phenotype', 'Tissue', 'Array', 'Ancestry', 'Developmental_period', 'Covariates', 'MPS_id']\n"
     ]
    }
   ],
   "source": [
    "# Merge BASE and TARGET data ---------------------------------------------------------\n",
    "d2_vars = ['Identifier', # 'Identifier_base', 'Multiple_identifier_base'\n",
    "           'Category', 'Phenotype', \n",
    "           'Tissue',\n",
    "           'Array',\n",
    "           'Ancestry',\n",
    "           'Developmental_period', \n",
    "           'Covariates']\n",
    "\n",
    "d1_vars = ['Identifier_base', # 'Multiple_identifier_base' ??\n",
    "           'Title', 'Year', 'What_is_available', 'Sample_overlap_target_base']+d2_vars[1:]\n",
    "\n",
    "# There is not overlap between BASE reference \n",
    "# any([i in base_va.Identifier for i in  base_ss.Identifier])\n",
    "d_base = pd.concat([base_ss[d2_vars], base_va[d2_vars]], axis=0)\n",
    "d_base['MPS_id'] = d_base['Identifier'] + d_base['Phenotype']\n",
    "print(d_base.shape, '\\n', list(d_base.columns))\n",
    "\n",
    "d_targ = lit.loc[lit.What_is_available != 'Only phenotype', d1_vars]\n",
    "d_targ['MPS_id'] = d_targ['Identifier_base'] + d_targ['Phenotype']\n",
    "d_targ = d_targ.rename(columns={'Identifier_base': 'Identifier'})\n",
    "\n",
    "print(d_targ.shape, '\\n', list(d_targ.columns))\n",
    "\n",
    "# Merge info about target with info about base (for summary statistics)\n",
    "d = d_targ.merge(d_base, on='MPS_id',\n",
    "                 # left_on='Identifier_base', right_on='Identifier',\n",
    "                 how='left', suffixes=['_targ','_base'])\n",
    "\n",
    "# TMP: remove mismatches -------------------------------------------------------------\n",
    "mismatch = d.loc[d['Identifier_base'].isna(), ]\n",
    "d = d.loc[d['Identifier_base'].notna(), ] # Effectively inner join...\n",
    "d.shape\n",
    "\n",
    "\n",
    "# Save cleaned data ------------------------------------------------------------------\n",
    "d.to_csv(f'{assets_directory}MPS_base_target.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c5453078-5cf4-4e5f-964b-586f2ca811ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Multiple_identifier_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Blood-based DNA methylation study of alcohol c...</td>\n",
       "      <td>Liu (2018) - A DNA methylation biomarker of, D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>Nasal DNA methylation at three CpG sites predi...</td>\n",
       "      <td>Xu (2021) - Shared DNA methylation signatures ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Title  \\\n",
       "81   Blood-based DNA methylation study of alcohol c...   \n",
       "434  Nasal DNA methylation at three CpG sites predi...   \n",
       "435  Nasal DNA methylation at three CpG sites predi...   \n",
       "436  Nasal DNA methylation at three CpG sites predi...   \n",
       "437  Nasal DNA methylation at three CpG sites predi...   \n",
       "438  Nasal DNA methylation at three CpG sites predi...   \n",
       "439  Nasal DNA methylation at three CpG sites predi...   \n",
       "\n",
       "                              Multiple_identifier_base  \n",
       "81   Liu (2018) - A DNA methylation biomarker of, D...  \n",
       "434  Xu (2021) - Shared DNA methylation signatures ...  \n",
       "435  Xu (2021) - Shared DNA methylation signatures ...  \n",
       "436  Xu (2021) - Shared DNA methylation signatures ...  \n",
       "437  Xu (2021) - Shared DNA methylation signatures ...  \n",
       "438  Xu (2021) - Shared DNA methylation signatures ...  \n",
       "439  Xu (2021) - Shared DNA methylation signatures ...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PROBLEMS \n",
    "# \"Identifier_base\" in the validated_algorithm - ignore? \n",
    "\n",
    "# What to do with the \"multiple\" bases ?  Excluded for now \n",
    "lit.loc[lit.Identifier_base == 'Multiple', ['Title', 'Multiple_identifier_base']]\n",
    "\n",
    "# mismatch\n",
    "\n",
    "# 21 mismatches... dropping them for now \n",
    "# Drop mismatches for now-- discuss with Is how to handle them \n",
    "# Richmond (2018) - DNA methylation as a marker --- does not match becasue two kinds of maternal smoking ...\n",
    "# Gondolia (2019) - Methylome-wide association study provides --- not match because Air Polution vs PM \n",
    "# Stevenson (2021) - Creating and Validating a DNA --- IL6 vs IL-6: CHANGED IT \n",
    "# Levy (2022) - Novel diagnostic DNA methylation --- different disorders mismatch...?\n",
    "# Awamleh (2022) - DNA methylation signature ass --- different syndromes?\n",
    "# Portales-Casamar (2016) - DNA methylation sign --- fetal alcohol syndrome ONLY? or also prenatal alcohol exposure \n",
    "# Aref-Eshghi (2019) - Evaluation of DNA Methyla --- Genitopatellar syndrome (GTPTS) and Ohdo syndrome, SBBYSS variant (SBBYSS)...  matching wich?\n",
    "# ...\n",
    "\n",
    "# Identifier base == multiple. What do you mean, combined? "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "demetra_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
